core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -740,6 +743,7 @@ private[spark] class TaskSetManager(
     }
     removeRunningTask(tid)
     info.markFinished(state)
+    successfulTaskDurations.remove(taskInfos(tid).duration)
```

Comment:
When we have executor failures (or due to pre-emption), this can essentially cause a large number of sub-optimal (new) speculative tasks to be executed.
Handling tasks which had not succeeded can be handled by looking at taskInfo to see if it has completed already ? (Yes, an if condition is required before remove).